### PR TITLE
add refresh sites endpoint

### PR DESCRIPTION
**Pull request by Ian-Nara**
[UID2-1069](https://atlassian.thetradedesk.com/jira/browse/UID2-1069)

### CODE REVIEW

File: src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        fakeAuth(OperatorType.PUBLIC, 99);
+        get(vertx, "/sites/refresh", "", ar -> {
+            assertTrue(ar.succeeded());
+            HttpResponse response = ar.result();
+            assertEquals(200, response.statusCode());
+            JsonObject json = response.bodyAsJsonObject();
+            String resultLocation = json.getJsonObject("sites").getString("location");
+            assertEquals(resultLocation, location);
+            testContext.completeNow();
+        });
```
Comment (mcollins-ttd) [inline src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java:106]: [Asserting in async code](https://vertx.io/docs/vertx-junit5/java/#_use_any_assertion_library) requires `verify`, `succeeding`, or `failing`.

```suggestion
        get(vertx, "/sites/refresh", "", testContext.succeeding(response -> testContext.verify(() -> {
            assertEquals(200, response.statusCode());
            JsonObject json = response.bodyAsJsonObject();
            String resultLocation = json.getJsonObject("sites").getString("location");
            assertEquals(resultLocation, location);
            testContext.completeNow();
        })));
```

File: src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        when(cloudStorage.download(eq(metadata))).thenReturn(new StringBufferInputStream(metadataContent));
+        when(cloudStorage.preSignUrl(any())).thenAnswer(i -> new URL(i.getArgument(0)));
+        fakeAuth(OperatorType.PRIVATE, 99);
+        get(vertx, "/sites/refresh", "", ar -> {
+            assertTrue(ar.succeeded());
+            HttpResponse response = ar.result();
+            assertEquals(400, response.statusCode());
+            assertEquals("endpoint /sites/refresh is for public operators only", response.bodyAsJsonObject().getString("message"));
+            testContext.completeNow();
+        });
```
Comment (mcollins-ttd) [inline src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java:123]:
```suggestion
        get(vertx, "/sites/refresh", "", testContext.succeeding(response -> testContext.verify(() -> {
            assertEquals(400, response.statusCode());
            assertEquals("endpoint /sites/refresh is for public operators only", response.bodyAsJsonObject().getString("message"));
            testContext.completeNow();
        })));
```

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java [hunk 3]
```diff
@@ -303,6 +306,21 @@ private static void setAttestationFailureReason(RoutingContext context, Attestat
         context.put(com.uid2.core.Const.RoutingContextData.ATTESTATION_FAILURE_DATA_PROP, data);
     }
 
+    private void handleSiteRefresh(RoutingContext rc) {
+        try {
+            OperatorInfo info = OperatorInfo.getOperatorInfo(rc);
+            if (info.getOperatorType() != OperatorType.PUBLIC) {
+                Error("error", 400, rc, "endpoint /sites/refresh is for public operators only");
```
Comment (mcollins-ttd) [inline src/main/java/com/uid2/core/vertx/CoreVerticle.java:313]: suggestion: 403 Forbidden instead of 400 Bad Request.

Comment (Ian-Nara) [reply src/main/java/com/uid2/core/vertx/CoreVerticle.java]: fixed this and updated in TestSiteSpecificMetadataPath.java as well

File: src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java (hunk 2 above)
Comment (Ian-Nara) [reply src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java]: fixed this and updated in TestSiteSpecificMetadataPath.java as well

File: src/main/java/com/uid2/core/vertx/CoreVerticle.java [hunk 4]
```diff
@@ -303,6 +306,21 @@ private static void setAttestationFailureReason(RoutingContext context, Attestat
...
+                return;
+            }
+            rc.response().putHeader(HttpHeaders.CONTENT_TYPE, "application/json")
+                    .end(siteMetadataProvider.getMetadata());
+        } catch (Exception e) {
+            logger.warn("exception in handleSiteRefresh: " + e.getMessage(), e);
+            Error("error", 500, rc, "error processing sties refresh");
```
[the remark this answers is not in the document]
Comment (Ian-Nara) [reply src/main/java/com/uid2/core/vertx/CoreVerticle.java]: fixed

File: src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    private JwtService jwtService;
+
+    // we need trusted to skip the attestation procedure or otherwise the core encpoint call made in this file will
+    // fail at the attestation handler
+    private static final String attestationProtocol = "trusted";
+
+    @BeforeEach
+    void deployVerticle(Vertx vertx, VertxTestContext testContext) throws Throwable {
+        attestationService = new AttestationService();
+        SecretStore.Global.load(((JsonObject) Json.decodeValue(openFile("/com.uid2.core/testSiteSpecificMetadata/test-secrets.json"))));
```
Comment (sunnywu) [inline src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java:69]: it took me a while to remember my work in site specific data stuff - for Site related tests here, they are not SiteSpecific as it's only available for public operator so could we setup a separate directory equivalent to testSiteSpecificMetadata for files required in this non-site-specific related tests here?

Might wanna change same paths for TestClientSideKeypairMetadataPath file too.

File: src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            assertEquals(200, response.statusCode());
+            JsonObject json = response.bodyAsJsonObject();
+            String resultLocation = json.getJsonObject("sites").getString("location");
+            assertEquals(resultLocation, location);
+            testContext.completeNow();
+        })));
+    }
+
+    @Test
+    void privateOperatorGetsKeypairsError(Vertx vertx, VertxTestContext testContext) throws CloudStorageException, IOException {
```
[the remark this answers is not in the document]
Comment (Ian-Nara) [reply src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java]: fixed

File: src/main/java/com/uid2/core/service/SiteMetadataProvider.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        String pathname = new GlobalScope(new CloudPath(SecretStore.Global.get(SiteMetadataPathName))).getMetadataPath().toString();
+        String original = readToEndAsString(metadataStreamProvider.download(pathname));
+        JsonObject main = (JsonObject) Json.decodeValue(original);
+        JsonObject obj = main.getJsonObject("sites");
+        String location = obj.getString("location");
+        obj.put("location", downloadUrlGenerator.preSignUrl(location).toString());
+        return main.encode();
+    }
+
+    private static String readToEndAsString(InputStream stream) throws IOException {
```
[the remark this answers is not in the document]
Comment (Ian-Nara) [reply src/main/java/com/uid2/core/service/SiteMetadataProvider.java]: oh oops, switching to Util

File: src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java (hunk 5 above)
Comment (Ian-Nara) [reply src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java]: refactored the test directorys/configs

File: src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+import static com.uid2.core.util.MetadataHelper.readToEndAsString;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(VertxExtension.class)
+public class TestSitesMetadataPath {
```
[the remark this answers is not in the document]
Comment (Ian-Nara) [reply src/test/java/com/uid2/core/vertx/TestSitesMetadataPath.java:46]: added comments

**Review by sunnywu — APPROVED** [review]

**Pull request merged**